### ISSUE TITLE
Run test Assemblies in parallel

### ISSUE THREAD

**Issue by CharliePoole** [nunit/nunit#304]
Implement parallel execution of test assemblies in the engine, as detailed in the spec at https://github.com/nunit/dev/wiki/Engine-Parallel-Test-Execution

Originally at https://github.com/nunit/nunit-console/issues/7

**Comment by CharliePoole** [comment]
@chris-smith-zocdoc Once you accept my invitation to the Contributors team I'll be able to assign this to you.

**Comment by CharliePoole** [comment]
Several different threads of activity currently pertain to this issue. I'll document them here:

@chris-smith-zocdoc Will produce a separate PR that makes TestRun an awaitable object and makes use of it throughout the hierarchy of tests.

@charliepoole will take PR #572 and extract a minimum feature set that implements parallel execution of assemblies, without the code that is intended to limit the number of assemblies. This is the feature currently desired for use with the VS adapter.

@OsirisTerje will introduce first execution in separate processes and then (when charlie's work is merged) parallel processes to the core engine, which currently excludes process execution.

Once all of the above are done, we'll look at designing implementing a method of throttling the number of parallel processes used.

**Comment by chris-smith-zocdoc** [comment]
@CharliePoole How far did you get in extracting the minimum feature set for parallel execution? Anything I can help out with?

**Comment by CharliePoole** [comment]
@chris-smith-zocdoc Just getting back to this after focusing on other stuff for a while. I'll let you know.

**Comment by CharliePoole** [comment]
Well, I just spent a bunch of time trying to make the approach I wanted to use - based on synchronization primitives - work across processes. It works fine otherwise, but cross-process is exactly the case we need. I can see a way to make it work across processes, but that will burden the non-cross-process case with a lot of overhead.

@chris-smith-zocdoc It turns out that the approach you originally took  using a separate thread to launch each parallel process was the best one. I'm retrofitting code based on your old branch to the current master. Not sure whether we should :-) or :-( about this one.

**Comment by chris-smith-zocdoc** [comment]
@CharliePoole :( Well that's unfortunate. Could you explain the cross-process issue you ran into and what your potential solution would be?

Do you have a branch that exhibits the cross-process issue? I wouldn't mind playing around with it and trying to find an alternative to make it work.

**Comment by CharliePoole** [comment]
The issue is that ManualResetEvent doesn't work cross process. I could make it a named EventWaitHandler instead, and put the name into the AsyncTestResult but that would be overkill for the inprocess waits.

Nope. I think you had the right approach at the start and that's what I used, making other changes but copying a lot of your code. There's now a PR waiting for review so you can take a look.
